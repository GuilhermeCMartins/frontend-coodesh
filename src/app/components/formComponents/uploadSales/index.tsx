import React, { useState } from "react";
import {
  Box,
  Typography,
  Input,
  Button,
  CircularProgress,
  Dialog,
  DialogTitle,
  DialogContent,
} from "@mui/material";
import { toast } from "react-toastify";
import { useSales } from "@/app/hooks/useSales";

interface UploadSalesProps {
  onClose: () => void;
  onUploadSuccess: () => void;
  open: boolean;
}

const UploadSales: React.FC<UploadSalesProps> = ({
  onClose,
  open,
  onUploadSuccess,
}) => {
  const [selectedFile, setSelectedFile] = useState<File | null>(null);
  const [uploading, setUploading] = useState(false);
  const { postSales } = useSales();

  const handleFileSelect = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.files && event.target.files.length > 0) {
      const file = event.target.files[0];
      setSelectedFile(file);
    }
  };

  const handleFileUpload = async () => {
    if (selectedFile) {
      setUploading(true);
      try {
        await postSales(selectedFile);
        toast.success("Arquivo de vendas enviado com sucesso!");
        onUploadSuccess();
      } catch (error) {
        toast.error(
          "Erro ao enviar arquivo de vendas. Por favor, tente novamente."
        );
      }
      setUploading(false);
      onClose();
    } else {
      toast.warning("Por favor, selecione um arquivo antes de fazer o upload.");
    }
  };

  return (
    <Dialog open={open} onClose={onClose}>
      <DialogTitle>Upload de Vendas</DialogTitle>
      <DialogContent>
        <Box p={3}>
          <Typography variant="h4" gutterBottom>
            Upload de Vendas
          </Typography>
          <Box display="flex" alignItems="center" marginBottom={2}>
            <Input
              type="file"
              inputProps={{ accept: ".csv, .xlsx, .xls, .txt" }}
              onChange={handleFileSelect}
            />
            <Box marginLeft={2}>
              <Button
                variant="contained"
                onClick={handleFileUpload}
                disabled={uploading}
              >
                {uploading ? <CircularProgress size={24} /> : "Enviar Arquivo"}
              </Button>
            </Box>
          </Box>
        </Box>
      </DialogContent>
    </Dialog>
  );
};

export default UploadSales;
