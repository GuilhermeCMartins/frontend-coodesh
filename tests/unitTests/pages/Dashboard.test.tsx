import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import DashboardPage from "../../../src/app/dashboard/page";
import { mockTransactions } from "./mocks/mockTransactions";
import { UploadSalesProps } from "@/app/components/formComponents/uploadSales";

jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
}));

jest.mock("../../../src/app/HOC/withAuth.tsx", () => {
  return (Component: React.ComponentType<any>) => {
    const WrappedComponent: React.FC<any> = (props) => <Component {...props} />;
    WrappedComponent.displayName = `withAuthentication(${
      Component.displayName || Component.name || "Component"
    })`;
    return WrappedComponent;
  };
});

jest.mock("../../../src/app/hooks/useTransactions", () => ({
  useTransactions: () => ({
    getAllTransactions: jest.fn().mockResolvedValue({
      transactions: mockTransactions,
    }),
  }),
}));

jest.mock("../../../src/app/components/formComponents/uploadSales", () => ({
  __esModule: true,
  default: ({ open, onClose }: UploadSalesProps) => (
    <div data-testid="upload-sales-dialog">
      UploadSales Dialog Content
      <button onClick={onClose} data-testid="close-button">
        Close
      </button>
    </div>
  ),
}));

describe("DashboardPage", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<DashboardPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the dashboard with transaction data", async () => {
    render(<DashboardPage />);

    await waitFor(
      () => expect(screen.queryByText("Carregando...")).not.toBeInTheDocument(),
      {
        timeout: 5000,
        interval: 100,
      }
    );

    await waitFor(async () => {
      const vendorNameElement = await screen.findByTestId("vendor-name");
      const productNameElement = await screen.findByTestId("product-name");
      const productPriceElement = await screen.findByTestId("product-price");

      expect(vendorNameElement.textContent).toBe("Vendor 1");
      expect(productNameElement.textContent).toBe("Product 1");
      expect(productPriceElement.textContent).toBe("R$10.00");
    });
  });

  it("applies filters correctly", async () => {
    render(<DashboardPage />);

    await waitFor(
      () => expect(screen.queryByText("Carregando...")).not.toBeInTheDocument(),
      {
        timeout: 5000,
        interval: 100,
      }
    );

    const vendorFilterInput = screen.getByRole("textbox", {
      name: /Filtrar por Vendedor/i,
    });
    const productFilterInput = screen.getByRole("textbox", {
      name: /Filtrar por Produto/i,
    });
    const minPriceFilterInput = screen.getByRole("textbox", {
      name: /Preço Mínimo/i,
    });

    fireEvent.change(vendorFilterInput, { target: { value: "Vendor 1" } });
    fireEvent.change(productFilterInput, { target: { value: "Product 1" } });
    fireEvent.change(minPriceFilterInput, { target: { value: "5.00" } });

    const applyFiltersButton = screen.getByRole("button", {
      name: /Aplicar Filtros/i,
    });
    fireEvent.click(applyFiltersButton);

    await waitFor(async () => {
      const vendorNameElement = await screen.findByTestId("vendor-name");
      const productNameElement = await screen.findByTestId("product-name");
      const productPriceElement = await screen.findByTestId("product-price");

      expect(vendorNameElement.textContent).toBe("Vendor 1");
      expect(productNameElement.textContent).toBe("Product 1");
      expect(productPriceElement.textContent).toBe("R$10.00");
    });
  });

  it("opens UploadSales dialog", async () => {
    render(<DashboardPage />);
    const openUploadSalesButton = screen.getByText("Upload transações");

    fireEvent.click(openUploadSalesButton);

    const uploadSalesDialogContent = screen.getByText(
      "UploadSales Dialog Content"
    );

    expect(uploadSalesDialogContent).toBeInTheDocument();
  });

  it("calculates total transactions correctly", async () => {
    render(<DashboardPage />);

    await screen.findByText("TRANSAÇÕES");

    await waitFor(
      () => expect(screen.queryByText("Carregando...")).not.toBeInTheDocument(),
      {
        timeout: 5000,
        interval: 100,
      }
    );

    const totalInbound = mockTransactions.reduce((total, transaction) => {
      return transaction.TransactionType.Inbound
        ? total + parseFloat(transaction.Price)
        : total;
    }, 0);

    const totalOutbound = mockTransactions.reduce((total, transaction) => {
      return !transaction.TransactionType.Inbound
        ? total + parseFloat(transaction.Price)
        : total;
    }, 0);

    const totalInboundElement = screen.getByTestId("inbound");
    const totalOutboundElement = screen.getByTestId("outbound");

    expect(totalInboundElement.textContent).toBe(
      `Entrada: R$${totalInbound.toFixed(2)}`
    );
    expect(totalOutboundElement.textContent).toBe(
      `Saída: R$${totalOutbound.toFixed(2)}`
    );
  });
});
