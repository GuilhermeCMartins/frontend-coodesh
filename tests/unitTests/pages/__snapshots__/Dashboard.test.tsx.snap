// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <section
      class="c-hITQoU c-hITQoU-bYMNqz-type-navbar"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="HomeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
        </span>
        Home
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Account"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeLarge css-1pux6rn-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </section>
    <section
      class="c-hITQoU c-hITQoU-jKOUQW-type-filters"
    >
      <div
        class="MuiBox-root css-19kzrtu"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        >
          FILTROS
        </h6>
        <div
          class="MuiBox-root css-9nrda2"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            data-testid="vendor-filter-input"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r1:"
              id=":r1:-label"
            >
              Filtrar por Vendedor
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Filtrar por Vendedor
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            data-testid="product-filter-input"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r2:"
              id=":r2:-label"
            >
              Filtrar por Produto
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r2:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Filtrar por Produto
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            data-testid="min-price-filter-input"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r3:"
              id=":r3:-label"
            >
              Preço Mínimo
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r3:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Preço Mínimo
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiBox-root css-d0uhtl"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              data-testid="apply-filters-button"
              tabindex="0"
              type="button"
            >
              Aplicar Filtros
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiBox-root css-d0uhtl"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Limpar Filtros
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </section>
    <section
      class="c-hITQoU c-hITQoU-dhzjXW-type-transactions"
    >
      <div
        class="MuiBox-root css-19kzrtu"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        >
          TRANSAÇÕES
        </h6>
        <div
          class="MuiBox-root css-0"
          style="width: 30vw; display: flex; justify-content: center; align-items: center;"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
          >
            Carregando...
          </p>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-middle MuiDivider-vertical MuiDivider-flexItem css-16bghm4-MuiDivider-root"
      />
      <div
        class="MuiBox-root css-1wxg2s5"
      >
        <div
          class="MuiBox-root css-0"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
          >
            Fazer uma nova transação:
          </h6>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Upload transações
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            data-testid="upload-sales-dialog"
          >
            UploadSales Dialog Content
            <button
              data-testid="close-button"
            >
              Close
            </button>
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
          >
            Total das transações:
          </h6>
          <div
            class="MuiBox-root css-0"
            data-testid="inbound"
          >
            Entrada: 
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-df5r72-MuiTypography-root"
            >
              R$0.00
            </h6>
          </div>
          <div
            class="MuiBox-root css-0"
            data-testid="outbound"
          >
            Saída: 
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-glud9z-MuiTypography-root"
            >
              R$0.00
            </h6>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
